eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/TxTimeLocksMonitor.scala

```diff
@@ -66,68 +66,103 @@ object TxTimeLocksMonitor {
+        Behaviors.same
+      case WrappedCurrentBlockHeight(_) => Behaviors.same
+      case CheckRelativeTimeLock => checkRelativeTimeLocks()
+    }
+  }
+
+  private def checkRelativeTimeLocks(): Behavior[Command] = {
     val csvTimeouts = Scripts.csvTimeouts(cmd.tx)
     if (csvTimeouts.nonEmpty) {
-      val watchConfirmedResponseMapper: ActorRef[WatchParentTxConfirmedTriggered] = context.messageAdapter(w => ParentTxConfirmed(w.tx.txid))
-      csvTimeouts.foreach {
+      val parentTxs = csvTimeouts.map {
         case (parentTxId, csvTimeout) =>
-          log.info("{} has a relative timeout of {} blocks, watching parentTxId={}", cmd.desc, csvTimeout, parentTxId)
-          watcher ! WatchParentTxConfirmed(watchConfirmedResponseMapper, parentTxId, minDepth = csvTimeout)
+          log.info("{} has a relative timeout of {} blocks, checking confirmations for parentTxId={}", cmd.desc, csvTimeout, parentTxId)
+          checkConfirmations(parentTxId)
+          parentTxId -> RelativeLockStatus(csvTimeout, nodeParams.currentBlockHeight + csvTimeout)
       }
-      waitForParentsToConfirm(csvTimeouts.keySet)
+      waitForRelativeTimeLocks(parentTxs)
     } else {
       notifySender()
     }
   }
 
-  def waitForParentsToConfirm(parentTxIds: Set[TxId]): Behavior[Command] = {
+  private case class RelativeLockStatus(csvTimeout: Long, checkAfterBlock: BlockHeight)
+
+  private def waitForRelativeTimeLocks(parentTxs: Map[TxId, RelativeLockStatus]): Behavior[Command] = {
     Behaviors.receiveMessagePartial {
-      case ParentTxConfirmed(parentTxId) =>
-        log.debug("parent tx of {} has been confirmed (parent txid={})", cmd.desc, parentTxId)
-        val remainingParentTxIds = parentTxIds - parentTxId
-        if (remainingParentTxIds.isEmpty) {
-          log.info("all parent txs of {} have been confirmed", cmd.desc)
-          notifySender()
-        } else {
-          log.debug("some parent txs of {} are still unconfirmed (parent txids={})", cmd.desc, remainingParentTxIds.mkString(","))
-          waitForParentsToConfirm(remainingParentTxIds)
+      case ParentTxStatus(parentTxId, confirmations_opt) =>
+        parentTxs.get(parentTxId) match {
+          case Some(status) => confirmations_opt match {
+            case Some(confirmations) if status.csvTimeout <= confirmations =>
+              log.debug("parentTxId={} of {} has reached enough confirmations", parentTxId, cmd.desc)
+              val remainingParentTxs = parentTxs - parentTxId
+              if (remainingParentTxs.isEmpty) {
+                log.info("all parent txs of {} have been confirmed", cmd.desc)
+                notifySender()
+              } else {
+                log.debug("some parent txs of {} are still unconfirmed (parentTxIds={})", cmd.desc, remainingParentTxs.keySet.mkString(","))
+                waitForRelativeTimeLocks(remainingParentTxs)
```

Comment:
I think you could probably do the same optimization here than below (i.e. skipping blocks for retrying). If the commit is unconfirmed and the delay is 2016 blocks, we could wait 2016 blocks before checking for confirmation? Or at least, maybe not check every block.

Reply:
🤐